br/pkg/restore/data.go

```diff
@@ -0,0 +1,507 @@
+// Copyright 2022 PingCAP, Inc. Licensed under Apache-2.0.
+package restore
+
+import (
+	"context"
+	"crypto/tls"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/emirpasic/gods/maps/treemap"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	recovpb "github.com/pingcap/kvproto/pkg/recoverdatapb"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/br/pkg/utils"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+// recover the tikv cluster
+// 1. read all meta data from tikvs
+// 2. Assemble region meta and make recovery plan
+// 3. send the recover plan and the wait tikv to apply, in waitapply, all assigned region leader will check and apply index to the last index
+// 4. send the resolved_ts to tikv for deleting data.
+func RecoverData(ctx context.Context, resolvedTs uint64, allStores []*metapb.Store, tls *tls.Config, dumpRegionInfo bool) (int, error) {
+
+	numOfTiKVs := len(allStores)
+	var recovery = NewRecovery(numOfTiKVs, tls)
+
+	err := recovery.ReadRegionMeta(ctx, numOfTiKVs, allStores)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	if dumpRegionInfo {
+		recovery.dumpRegionInfo()
+	}
+
+	totalRegions := recovery.getTotalRegions()
+
+	recovery.makeRecoveryPlan()
+
+	err = recovery.RecoverRegions(ctx, allStores)
+	if err != nil {
+		return totalRegions, errors.Trace(err)
+	}
+
+	// TODO Sleep is workaround for sync last index, it will be removed when TiKV part of work been done
+	time.Sleep(time.Second * 3)
+
+	err = recovery.ResolveData(ctx, allStores, resolvedTs)
+	if err != nil {
+		return totalRegions, errors.Trace(err)
+	}
+	return totalRegions, nil
+}
+
+type RecoveryMeta struct {
+	storeId      uint64
+	recoveryMeta []*recovpb.RegionMeta
+}
+
+func NewRecoveryMeta(storeId uint64) RecoveryMeta {
+	var meta = make([]*recovpb.RegionMeta, 0)
+	return RecoveryMeta{storeId, meta}
+}
+
+type RecoveryPlan struct {
+	storeId      uint64
+	recoveryPlan []*recovpb.RecoverCmdRequest
+}
+
+func NewRecoveryPlan(storeId uint64) RecoveryPlan {
+	var meta = make([]*recovpb.RecoverCmdRequest, 0)
+	return RecoveryPlan{storeId, meta}
+}
+
+type Recovery struct {
+	instances    int
+	instanceId   *uint32
+	regionMetas  []RecoveryMeta
+	recoveryPlan map[uint64][]*recovpb.RecoverCmdRequest
+	resolvedTs   *uint64
```

Comment:
no need to use pointer

Reply:
removed instanceId, and others for gRPC used.